backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/questionsubmission/QuestionSubmissionService.java

```diff
@@ -0,0 +1,278 @@
+    }
+
+    @Retryable(value = {SQLException.class}, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void removeReviews(QuestionSubmission questionSubmission) {
+        List<Review> reviews = new ArrayList<>(reviewRepository.findQuestionSubmissionReviews(questionSubmission.getId()));
+        for (Review review : reviews) {
+            review.remove();
+            reviewRepository.delete(review);
+        }
+    }
+
+    @Retryable(value = {SQLException.class}, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void toggleInReviewStatus(int questionSubmissionId, boolean inReview) {
+        QuestionSubmission questionSubmission = getQuestionSubmission(questionSubmissionId);
+        Review.Status status = inReview ? Review.Status.IN_REVIEW : Review.Status.IN_REVISION;
+        updateQuestionStatus(status, questionSubmission.getQuestion().getId());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<QuestionSubmissionDto> getStudentQuestionSubmissions(Integer studentId, Integer courseExecutionId) {
+        return questionSubmissionRepository.findQuestionSubmissions(studentId, courseExecutionId).stream().map(QuestionSubmissionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<QuestionSubmissionDto> getCourseExecutionQuestionSubmissions(Integer courseExecutionId) {
+        return questionSubmissionRepository.findCourseExecutionQuestionSubmissions(courseExecutionId).stream().map(QuestionSubmissionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<ReviewDto> getQuestionSubmissionReviews(Integer questionSubmissionId) {
+        return reviewRepository.findQuestionSubmissionReviews(questionSubmissionId).stream().map(ReviewDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<UserQuestionSubmissionInfoDto> getAllStudentsQuestionSubmissionsInfo(Integer courseExecutionId) {
+        CourseExecution courseExecution = getCourseExecution(courseExecutionId);
+        Set<User> students = courseExecution.getStudents();
+
+        List<UserQuestionSubmissionInfoDto> userQuestionSubmissionInfoDtos = new ArrayList<>();
+
+        for (User student: students) {
+            List<QuestionSubmissionDto> questionSubmissions = student.getQuestionSubmissions().stream().map(QuestionSubmissionDto::new).collect(Collectors.toList());
```

Comment:
this can be inside the constructor of UserQuestionSubmissionInfoDto, it receives a student and copies its question submissions